### PR TITLE
adds different scopes for datadirs

### PR DESCRIPTION
**Pull request by salotz**
In order to accomplish this I generalized the creation of datadirs,
and then provided aliases for the old fixtures of the same name.

This adds a number of fixtures to the namespace which are the same as
before but prefixed by the pytest scope they apply to,
e.g. `function_shared_datadir`.  Which is the same as the old
`shared_datadir`.

This also opens up for more complex (i.e. non-hardcoded) usage
scenarios, although I have just stuck to the same behaviors as before.

The only difference perhaps is that previously there was a behavior
where if the original data dir was not a directory at all a directory
in the temp dir was created with that name, although I didn't quite
understand why that was needed or what pattern it supports.

### CODE REVIEW

**Comment by salotz** [comment]
Fixes #26 (even though it was closed). Let me know what you think. Feel free to make suggestions and edits.

**Comment by salotz** [comment]
I'll work on fixing the tests, just wanted to get this on here to see if there was any big objections to a change like this.

**Comment by igortg** [comment]
Seems a very nice addition. If there is no backward compatibility, it's a go for me.

**Comment by salotz** [comment]
Realized there was an issue with the previous implementation. Shouldn't have been so quick to do metaprogramming. New one is much cleaner IMO, although not parametric in the scope nor does it work for sessions (which is a limitation of `tmpdir` and shouldn't have worked last time either).

File: pytest_datadir/plugin.py [hunk 1]
```diff
@@ -25,24 +26,106 @@ def _win32_longpath(path):
...
-    return result
+class DatadirFactory(object):
+    """Factory class for generating datadir fixtures."""
+
+    def __init__(self, request, tmp_path_factory):
+
+        self.tmp_path_factory = tmp_path_factory
+        self.request = request
+
+    def mkdatadir(self, original_datadir='data'):
```
Comment (nicoddemus) [inline pytest_datadir/plugin.py]: Hmm seems strange to me to have `'data'` as default. I would expect this to be `None`. Do you have a use-case?

File: pytest_datadir/plugin.py [hunk 2]
```diff
@@ -25,24 +26,106 @@ def _win32_longpath(path):
...
+        # dir
+        shutil.copytree(original_path, temp_data_path)
+
+        return temp_data_path
+
+# scoped factories
+@pytest.fixture(scope='module')
+def module_datadir_factory(request, tmp_path_factory):
+
+    return DatadirFactory(request, tmp_path_factory)
```
Comment (nicoddemus) [inline pytest_datadir/plugin.py]: `tmp_path_factory` is session-scoped, and all fixtures are using it even though they have different scopes. This is fine, but this means the directories won't be removed at the end of their respective scopes (`module_datadir_factory` for example won't remove its directory at the end of its module scope).

Is that what you intended?

File: pytest_datadir/plugin.py (hunk 1 above)
Comment (salotz) [reply pytest_datadir/plugin.py]: My thought process: I was thinking that since this was a hardcoded behavior in the past that if someone were to want to replicate that behavior they would have to look up the specific string for backwards compatibility, so I thought it was a good place to put it (there should also be a module or class constant as well). I view `None` (and secondarily `Ellipsis`) as sort of the other obvious defaults since they only signify themselves and would have to be interpreted by special logic in the module to mean anything. For strings the user is free to make their own labels however there is one semi-special string, 'data', that is a module specific convention and the user looking at the autodocs signature or docstring and seeing that as the default would recognize that.

I think your judgement is best since I am really brand new to this sphere.

File: pytest_datadir/plugin.py (hunk 2 above)
Comment (salotz) [reply pytest_datadir/plugin.py]: Yes I think this is an issue which could cause considerable confusion. What do you think of this: https://github.com/salotz/pytest-datadir/blob/scoped_tmp_path_factory/pytest_datadir/plugin.py#L81
Could almost be pushed upstream...

**Comment by salotz** [comment]
Does this behavior really make sense https://github.com/gabrielcnr/pytest-datadir/blob/master/tests/test_nonexistent.py ? I'm all for backwards compatibility, but this sounds like a silent failure. If you ask for a datadir but don't have one you probably want to remove that fixture from your test, right? If you just want a tmpdir, then just ask for that fixture.

**Comment by nicoddemus** [comment]
> Does this behavior really make sense /tests/test_nonexistent.py@master  ?

I understand the reasoning, but having the `datadir` work like `tmpdir` if there is no data directory is very convenient; it saves you from having to refactor in case things change. Please keep the existing behavior. 👍

File: pytest_datadir/plugin.py (hunk 2 above)
Comment (salotz) [reply pytest_datadir/plugin.py]: Merged that proposal for solving this.

File: pytest_datadir/plugin.py [hunk 3]
```diff
@@ -25,24 +30,149 @@ def _win32_longpath(path):
...
+
+@pytest.fixture(scope="module")
+def module_tmp_path_factory(request):
+    """Return a :class:`_pytest.tmpdir.TempPathFactory` instance for the test session.
+    """
+    return request.config._tmp_path_factory
+
+
+@pytest.fixture(scope="class")
+def class_tmp_path_factory(request):
```
Comment (nicoddemus) [inline pytest_datadir/plugin.py]: I don't think it makes sense to have class/module/function `tmp_path_factory` fixtures if they all return `tmp_path_factory`, which is session-scoped. Session-scoped fixtures can be used from any other scope after all. Also, `DatadirFactory` always copies from a directory named after the module name anyway (`self.request.module.__file__`).

File: pytest_datadir/plugin.py [hunk 4]
```diff
@@ -25,24 +30,149 @@ def _win32_longpath(path):
...
+
+@pytest.fixture(scope="function")
+def function_tmp_path_factory(request):
+    """Return a :class:`_pytest.tmpdir.TempPathFactory` instance for the test session.
+    """
+    return request.config._tmp_path_factory
+
+# scoped factories
+@pytest.fixture(scope='module')
+def module_datadir_factory(request, module_tmp_path_factory):
```
Comment (nicoddemus) [inline pytest_datadir/plugin.py]: Same reasoning here, it doesn't make sense to have multiple scopes for datadir factories, just a session-scoped `datadir_factory` would fit the bill nicely.

File: pytest_datadir/plugin.py [hunk 5]
```diff
@@ -25,24 +30,149 @@ def _win32_longpath(path):
...
+@pytest.fixture(scope='function')
+def function_datadir_factory(request, function_tmp_path_factory):
+
+    return DatadirFactory(request, function_tmp_path_factory)
+
+
+# shared datadirs
+
+@pytest.fixture(scope='module')
+def module_shared_datadir(request, module_datadir_factory):
```
Comment (nicoddemus) [inline pytest_datadir/plugin.py]: I'm not sure having specific "shared_datadir" fixtures for each scope makes sense. Users could just use `datadir_factory` directly, similar to how pytest does it.

File: pytest_datadir/plugin.py [hunk 6]
```diff
@@ -25,24 +30,149 @@ def _win32_longpath(path):
...
+
+    return function_datadir_factory.mkdatadir()
+
+shared_datadir = function_shared_datadir
+
+
+# test module datadirs
+
+@pytest.fixture(scope='module')
+def module_datadir(request, module_datadir_factory):
```
Comment (nicoddemus) [inline pytest_datadir/plugin.py]: Ditto here about having multiple scopes.

**Comment by salotz** [comment]
Hi @nicoddemus,

I really appreciate your patience and taking the time to review these and give me feedback and actually understanding my intention. I will cogitate on this a bit to make sure I understand it, as you pointed out some errors in correctness of the code's intended behavior.

My motivation for the for having datadir fixtures at different scopes was as follows:

1. The "function" scoped tmpdir was really an implicit scope. I like and tend to follow in my own codebases the "explicit is better than implicit" mantra. Despite the added verbosity (which I feel is more of a design problem with implicit in-band signalling with pytest fixture names) I think it is much more explicit about what you are doing. I had to go in and read the code to find out what scope tmpdir, datadir etc. were which is obfuscation IMO.

2. tmpdir and datadir fixtures at scopes other than session and function are useful. I like how you can rely on the scoping to do the teardown stuff for you so that you don't have to worry about making and naming separate directories for different tests, just because you have to use the session-scoped tmp_path_factory. This was one of the features that led me to adopting pytest in the first place. If they aren't in pytest-datadir I would probably end up adding them to my conftest.py or my own fork, but then this package would just be passing off the buck of boilerplate to users (unless I am the only one), which kind of defeats the purpose of a module. 

I don't want to start feature creep but the scope of functionality for this project was very small to begin with, and I hardly think that this increases the complexity of the codebase or API by much. 

In particular I would like to know anyone else's take on the issues in point 2. How do you handle, say, a datadir/tmpdir for a class scope?

Hopefully, I'm not making a mountain of a mole hill here, but these are motivated by my own real usage of the package and the boilerplate I wrote to address the packages shortcomings.

**Comment by nicoddemus** [comment]
> I really appreciate your patience and taking the time to review these and give me feedback and actually understanding my intention.

No problem at all, always glad to receive and discuss new contributions. 👍 

> I had to go in and read the code to find out what scope tmpdir, datadir etc. were which is obfuscation IMO.

Fair point, but this probably should be improved on pytest itself to mention the fixture scope with `pytest --fixtures`. 

> tmpdir and datadir fixtures at scopes other than session and function are useful. I like how you can rely on the scoping to do the teardown stuff for you so that you don't have to worry about making and naming separate directories for different tests

But please keep in mind that just by making a fixture module or class-scoped, that the directories will be removed automatically at the end of that scope. The only thing the fixture scope dictates is that the "teardown" code of the fixture will be executed when the fixture gets out of scope, but we don't have any specific teardown code in place that removes the directories created by the fixture: we rely on the `tmp_path_factory` behavior that removes directories from previous runs.

> In particular I would like to know anyone else's take on the issues in point 2. How do you handle, say, a datadir/tmpdir for a class scope?

I've never needed this with `datadir`, but the few times this has come up with `tmp_path` I just used `tmp_path_factory` to create a new directory. Note that I don't care that the directory effectively gets removed when the fixture gets out of scope, I delegate to the `tmp_path_factory` to remove the created directory eventually.

File: pytest_datadir/plugin.py (hunk 3 above)
Comment (salotz) [reply pytest_datadir/plugin.py]: In the code for `tmp_path_factory` that this is exactly how it is scoped as session: 
https://github.com/pytest-dev/pytest/blob/0b6960446edce271d0af6dff3293f7eb17f766c1/src/_pytest/tmpdir.py#L155

It returns a `TmpPathFactory` object from the config object which is scope-less and just runs `make_numbered_dir`. So the code here should be enough to change the scope of the fixture.

> Session-scoped fixtures can be used from any other scope after all.

This is the kind of shared state I am trying to avoid :P 

> Also, DatadirFactory always copies from a directory named after the module name anyway (self.request.module.__file__).

Strictly that isn't true, that only happens with the when `original_datadir is None`.

This point though is relevant if your datadir is read-only, in which case you would never need anything other than the session scoped fixture, since you don't really want to be copying multiple times (I am for that being the default behavior). I think there should be the ability to use them read-write and allow for isolation via scopes though.

So basically this is a workaround for the lack of intermediate scopes in `tmp_path_factory` and I am starting to think that this really should be a PR for the main pytest repo for `tmp_path_factory`.

Comment (nicoddemus) [reply pytest_datadir/plugin.py]: Just to make sure we are on the same page, let me ask you a question: when a module-scoped `datadir_factory` goes out of scope (IOW, we start running tests for another module), do you expect that the directories created by `datadir_factory` to be removed?

Comment (salotz) [reply pytest_datadir/plugin.py]: No, I just want a fresh context. So I want `make_numbered_dir` to be called and the basetemp to be set. If I modified the files in the datadir in one module and then need the same datadir for another module the modifications aren't still present.

Comment (nicoddemus) [reply pytest_datadir/plugin.py]: > No, I just want a fresh context.

Ahh I see, OK then. 👍 

Then overall your patch makes sense, I just don't think we need factories for each scope, or even shared-datadirs for each scope too. How about we have only datadir fixtures for each scope?

But perhaps we don't even need the datadir fixtures per-scope variants? I mean:

```python
@pytest.fixture(scope='module')
def mydata(module_datadir):
   store = module_datadir / 'store'
   ...
```

Can just as easily be written as:

```python
@pytest.fixture(scope='module')
def mydata(datadir_factory):
   store = datadir_factory.mkdatadir() / 'store'
   ...
```

Which is exactly the same number of lines of code. Having a small API is beneficial, both from a maintenance POV as well as usability.

@igortg @edisongustavo @tadeu what you guys think?

Comment (tadeu) [reply pytest_datadir/plugin.py]: I like the `datadir_factory` approach, looks more simple indeed

**Comment by jmehnle** [comment]
I was this was brought to a conclusion.

**Comment by salotz** [comment]
> I was this was brought to a conclusion.

For me in the places I needed it I just used my fork: https://github.com/salotz/pytest-datadir-extras
I haven't needed it in a long time though, hence the lack of activity. From this thread it seemed like it wasn't something they wanted to include, but it wasn't closed so I guess if it was matured they might still consider it.

**Comment by nicoddemus** [comment]
It is up to @gabrielcnr, but for my part I don't have interest on this feature, particularly liking that the plugin is lean and simple.

**Comment by tboddyspargo** [comment]
> The main idea is to reduce the amount of copying that is needed for large inputs. If you are using the same data over and over again for many tests calling mkdatadir() over and over will make your tests much slower than they need to be. The `module_datadir` etc. just copies once for the scope

I just stumbled across this and would like to echo this part of the motivation for this PR. I have a very large number of test cases and it seems that the implicit `function` scope of this fixture leads to a significant performance hit for what seems to be redundant `shutil.copytree` (copied all directory contents for every test case). My usage of `pytest-datadir` is indirect via the `pytest-regressions` plugin. Unfortunately, the performance impact is such that we have had to turn to other solutions.

**Comment by gabrielcnr** [comment]
I appreciate the request and can see how a feature like this could be desirable. I understand that the main motivation for this is performance, as explained before with the use case for "large files" which can lead to a sensible overhead during copies.

However I believe that this brings some complexity to the plugin, including its architectural foundation, which is not something that I believe would pay off, when balancing that against the main / most common use cases. 

It would be even more simple to expose something like a `source_datadir` which would give a Path to the original directory containing the datadir files (not the copy), which then would allow users to have more freedom to add specific behaviour they might wish/like, making it easy to address the case of the copy-performance-overhead problem by simply having some kind of caching/memoization that reads the desired (ultra large) file only once, on-demand.

I'm inclined to agree with @nicoddemus here and opt for keeping simplicity.

**Comment by nicoddemus** [comment]
Given the discussion here and in https://github.com/ESSS/pytest-regressions/issues/190, perhaps we can implementa *another* fixture in this plugin, that copy files lazily? I think this would solve the heart of the issue.

As discussed in https://github.com/ESSS/pytest-regressions/issues/190, we could have a new fixture, say `lazy_datadir` which is implemented by this object:

```python
@dataclass(frozen=True)
class LazyDataDir
    # The original data directory, kept under version control. 
    original_datadir: Path
    # Temporary directory, starts empty.
    tmp_path: Path 
    
    def joinpath(self, other: str) -> Path:
        original = self.original_datadir / other
        target = self.tmp_path / other
        if original.is_file():
            shutil.copy(original, target)
        return target

   def __div__(self, other: str) -> Path:
        return self.joinpath(other)      

@pytest.fixture
def lazy_datadir(original_datadir: Path, tmp_path: Path) -> LazyDataDir:
    return LazyDataDir(original_datadir, tmp_path)  
```

The usage would be thus:

```python
def test_foo(lazy_datadir: LazyDataDir) -> None:
    data_file = lazy_datadir / "foo.png"  # copies from the original datadir, if exists.

    new_file lazy_datadir.tmp_path / "new.png"    
``` 


This seems simple to implement and would cover the overall need for performance, including the one discussed in https://github.com/ESSS/pytest-regressions/issues/190, while being fully backward compatible.

**Comment by gabrielcnr** [comment]
Sounds good - certainly good addition

Please feel free to go ahead and open a PR with this sketch.

But, still, considering the use case of having _so many_ tests using `lazy_datadir` to copy the same _large file_ , on-demand, the performance problem would still exist if I'm not mistaken. Shouldn't we also give a trivial and straight access to the original datadir directory? That way the users can decide if they want to implement their own "session-scoped datadir" or have their own caching mechanisms in place, etc?

**Comment by nicoddemus** [comment]
> Please feel free to go ahead and open a PR with this sketch.

I will open an issue, but won't have time anytime soon to work on this myself.

> Shouldn't we also give a trivial and straight access to the original datadir directory? 

We already provide the [`original_datadir fixture`](https://github.com/gabrielcnr/pytest-datadir/blob/70edaa2b070ed9ddb8da4734c8d36557d7c7671e/src/pytest_datadir/plugin.py#L43-L45).

Closing this then.
